src/webgl.js

```diff
@@ -183,13 +183,15 @@ function initCopyProgram(gl) {
   layout(location = 0) out vec4 copyOut;
 
   void main() {
-    if (gl_FragCoord[0] < renderArea.x) {
-      copyOut = vec4(0.0, 0.0, 0.0, 1.0);
-    } else if (gl_FragCoord[0] > renderArea.z) {
-      copyOut = vec4(0.0, 0.0, 0.0, 1.0);
-    } else if ((gl_FragCoord[1] < renderArea.y) || (gl_FragCoord[1] > renderArea.w)) {
-      copyOut = vec4(0.0, 0.0, 0.0, 1.0);
-    } else {
+    // default return black
+    copyOut = vec4(0.0, 0.0, 0.0, 1.0);
+
+    // if we are not in the padding around the video, get the texture value
+    if (gl_FragCoord.x > renderArea.x
+      && gl_FragCoord[0] < renderArea.z
+      && gl_FragCoord[1] > renderArea.y
+      && gl_FragCoord[1] < renderArea.w
+    ) {
       copyOut = texture(originalSampler, vec2((gl_FragCoord[0] - renderArea.x) / videoRes.x, 1.0 - ((gl_FragCoord[1] - renderArea.y) / videoRes.y)));
```

Comment:
Also in line with pairwise optimization, I think you can do:

```
copyOut = texture(originalSampler, (gl_FragCoord.xy - renderArea.xy) / videoRes.xy);
```

Important GL insight here is that you can typically perform any operation you want to do on the components of a vector, directly on the vector; I think the operation is accelerated in hardware, if so it would actually make a substantial performance difference as well.

Reply:
OK, for educational purposes, I wanted to throw out another couple potential optimizations. Branches are typically more expensive than nonbranching code as I think you mentioned last week; the usual solution is to apply something like a step function to replace the condition. Here, I think that could look like (untested):

```
copyOut = vec4(0.0, 0.0, 0.0, 1.0);
vec2 mask = step(renderArea.xy, gl_FragCoord.xy) - step(renderArea.zw, gl_FragCoord.xy);
copyOut = mask.x * mask.y * texture(originalSampler, (gl_FragCoord.xy - renderArea.xy) / videoRes.xy);
```

The two calls to `step()` cause `mask` to be 0 up to the inset; 1 while over top of the inset; then 0 again after passing the inset. They are evaluated independently in each dimension yielding a `vec2` that is always either 0 or 1 in each axis. You then multiply them both by the texture value, because you only want to retain color where both axes are 1.

I wouldn't be surprised if you could go further than that, but I bet it isn't super duper worthwhile because of the size of the image while running this code.

Reply:
> Also in line with pairwise optimization, I think you can do:
> 
> ```
> copyOut = texture(originalSampler, (gl_FragCoord.xy - renderArea.xy) / videoRes.xy);
> ```
> 
> Important GL insight here is that you can typically perform any operation you want to do on the components of a vector, directly on the vector; I think the operation is accelerated in hardware, if so it would actually make a substantial performance difference as well.

I like the look of that better, much cleaner and easier to understand. But there's that 1.0 - for the .y part of the expression, is there an easy way to sneak that in? I tried with making a tmpVar, doing the math using the vectors, and then just doing a `tmpVar.y = 1.0 - tmpVar.y` but that ended up with more lines, and it's possible it tends to run just a bit slower than what we have currently.

```glsl
vec2 tmpVar;
tmpVar.xy = (gl_FragCoord.xy - renderArea.xy) / videoRes.xy;
tmpVar.y = 1.0 - tmpVar.y;
copyOut = texture(originalSampler, tmpVar);
```